File: src/agy_swap/quota.py
"""Google Cloud Code API quota fetching and batch background refresh."""

from concurrent.futures import ThreadPoolExecutor, as_completed
from datetime import datetime, timezone

from agy_swap import TIER_NAMES, QUOTA_CACHE_SECS, QuotaFetchError, AccountStoreError
from agy_swap.display import clean_display_text, _parse_utc_datetime
from agy_swap.oauth import _refresh_access_token, _cloud_code_post, extract_verified_google_email_claim
from agy_swap.store import _normalize_quota_snapshot, save_accounts


def fetch_account_quota(account):
    access_token = _refresh_access_token(account.get("token_data"))
    claimed_email = extract_verified_google_email_claim(account.get("token_data"))
    if claimed_email and claimed_email != account.get("email"):
        raise QuotaFetchError("refreshed token identity does not match the account")

    account_info = _cloud_code_post(access_token, "loadCodeAssist", {"metadata": {"ideType": "ANTIGRAVITY"}})
    project = account_info.get("cloudaicompanionProject")
    if not isinstance(project, str) or not project:
        raise QuotaFetchError("Google returned no Code Assist project")
    summary = _cloud_code_post(access_token, "retrieveUserQuotaSummary", {"project": project})

    tier = account_info.get("paidTier") or account_info.get("currentTier")
    if not isinstance(tier, dict) or not tier.get("id"):
        raise QuotaFetchError("Google returned no account tier")
    groups = []
    for raw_group in summary.get("groups", []):
        if not isinstance(raw_group, dict):
            continue
        buckets = []
        group_id = None
        for raw_bucket in raw_group.get("buckets", []):
            if not isinstance(raw_bucket, dict):
                continue
            bucket_id = clean_display_text(raw_bucket.get("bucketId"))
            window = raw_bucket.get("window")
            fraction = raw_bucket.get("remainingFraction")
            reset_at = raw_bucket.get("resetTime")
            if bucket_id.startswith("gemini-"):
                group_id = "gemini"
            elif bucket_id.startswith("3p-"):
                group_id = "third_party"
            if window not in ("weekly", "5h") or isinstance(fraction, bool) or not isinstance(fraction, (int, float)) or not isinstance(reset_at, str):
                continue
            buckets.append({
                "id": bucket_id,
                "name": clean_display_text(raw_bucket.get("displayName"), window),
                "window": window,
                "remaining_fraction": min(1.0, max(0.0, float(fraction))),
                "reset_at": reset_at,
            })
        if group_id and buckets:
            groups.append({
                "id": group_id,
                "name": clean_display_text(raw_group.get("displayName"), "Model group"),
                "buckets": buckets,
            })
    if not groups:
        raise QuotaFetchError("Google returned no quota groups")

    tier_id = clean_display_text(tier.get("id"))
    snapshot = {
        "observed_at": datetime.now(timezone.utc).isoformat(),
        "tier": {"id": tier_id, "name": TIER_NAMES.get(tier_id, clean_display_text(tier.get("name"), tier_id))},
        "groups": groups,
    }
    try:
        return _normalize_quota_snapshot(snapshot, account.get("email", "account"))
    except AccountStoreError:
        raise QuotaFetchError("Google returned invalid quota data") from None


def quota_snapshot_age(account, now=None):
    snapshot = account.get("quota_snapshot")
    if not isinstance(snapshot, dict):
        return None
    try:
        return max(0, ((now or datetime.now(timezone.utc)) - _parse_utc_datetime(snapshot["observed_at"])).total_seconds())
    except (KeyError, AttributeError, TypeError, ValueError):
        return None


def refresh_quota_snapshots(accounts, force=False, progress=None):
    errors = {}
    changed = False
    now = datetime.now(timezone.utc)
    total = len(accounts)

    to_fetch = []
    cached_count = 0
    for email, account in accounts.items():
        age = quota_snapshot_age(account, now)
        if not force and age is not None and age < QUOTA_CACHE_SECS:
            cached_count += 1
            if progress:
                progress(cached_count, total, account, "cached")
        else:
            to_fetch.append((email, account))

    if to_fetch:
        def _fetch_one(item):
            email, account = item
            try:
                snapshot = fetch_account_quota(account)
                return email, snapshot, None
            except QuotaFetchError as exc:
                return email, None, str(exc)

        workers = min(len(to_fetch), 8)
        with ThreadPoolExecutor(max_workers=workers) as executor:
            futures = {executor.submit(_fetch_one, item): item[0] for item in to_fetch}
            done_count = cached_count
            for future in as_completed(futures):
                email, snapshot, error = future.result()
                done_count += 1
                if snapshot:
                    accounts[email]["quota_snapshot"] = snapshot
                    changed = True
                    status = "synced"
                else:
                    errors[email] = error
                    status = "failed"
                if progress:
                    progress(done_count, total, accounts[email], status)

    if changed:
        save_accounts(accounts)
    return errors


def get_token_reset_info(token_data_str):
    from agy_swap.oauth import decode_token
    token_json = decode_token(token_data_str)
    if not token_json or "token" not in token_json:
        return None
    token_obj = token_json["token"]
    expiry_str = token_obj.get("expiry")
    has_refresh = bool(token_obj.get("refresh_token"))

    try:
        if expiry_str:
            dt = _parse_utc_datetime(expiry_str)
        elif token_obj.get("expiry_date"):
            dt = datetime.fromtimestamp(float(token_obj["expiry_date"]) / 1000, tz=timezone.utc)
        else:
            return None
        now = datetime.now(timezone.utc)
        diff = (dt - now).total_seconds()
        if diff > 0:
            mins = int(diff // 60)
            secs = int(diff % 60)
            if mins >= 60:
                hours = mins // 60
                rem_mins = mins % 60
                return {"status": "active", "remaining_str": f"{hours}h {rem_mins}m", "seconds": diff, "expiry_dt": dt}
            return {"status": "active", "remaining_str": f"{mins}m {secs}s", "seconds": diff, "expiry_dt": dt}
        else:
            if has_refresh:
                return {"status": "refresh_available", "remaining_str": "Access expired · refresh token available", "seconds": diff, "expiry_dt": dt}
            else:
                abs_mins = int(abs(diff) // 60)
                return {"status": "expired", "remaining_str": f"Expired {abs_mins}m ago", "seconds": diff, "expiry_dt": dt}
    except (TypeError, ValueError, OverflowError):
        return None

File: internal/app/quota.go
package app

import (
	"context"
	"fmt"
	"strings"
	"sync"
	"time"
)

type quotaProgress func(index, total int, account Account, state string)

type QuotaService struct {
	http  *HTTPService
	store *Store
	vault AccountVault
}

func NewQuotaService(httpService *HTTPService, store *Store) *QuotaService {
	return &QuotaService{http: httpService, store: store}
}

func (q *QuotaService) SetVault(vault AccountVault) { q.vault = vault }

func (q *QuotaService) Fetch(ctx context.Context, account Account) (map[string]any, error) {
	tokenData, err := accountToken(ctx, account, q.vault)
	if err != nil {
		return nil, err
	}
	access, refreshed, err := q.http.accessTokenData(ctx, tokenData)
	if err != nil {
		return nil, err
	}
	stored := tokenData
	if refreshed != tokenData {
		stored = refreshed
	}
	if !tokenMatchesEmail(stored, getString(account, "email")) {
		return nil, fmt.Errorf("refreshed token identity does not match the account")
	}
	if refreshed != tokenData {
		account["token_data"] = refreshed
		delete(account, "secret_ref")
		rememberTokenExpiry(account, refreshed)
	}
	info, err := q.http.cloudPost(ctx, access, "loadCodeAssist", map[string]any{"metadata": map[string]any{"ideType": "ANTIGRAVITY"}})
	if err != nil {
		return nil, err
	}
	project := getString(info, "cloudaicompanionProject")
	if project == "" {
		return nil, fmt.Errorf("Google returned no Code Assist project")
	}
	summary, err := q.http.cloudPost(ctx, access, "retrieveUserQuotaSummary", map[string]any{"project": project})
	if err != nil {
		return nil, err
	}
	tier := getMap(info["paidTier"])
	if tier == nil {
		tier = getMap(info["currentTier"])
	}
	tierID := cleanText(getString(tier, "id"))
	if tierID == "" {
		return nil, fmt.Errorf("Google returned no account tier")
	}
	tierName := tierNames[tierID]
	if tierName == "" {
		tierName = cleanText(firstString(tier["name"], tierID))
	}
	groups := make([]any, 0)
	for _, rawGroup := range getSlice(summary["groups"]) {
		group := getMap(rawGroup)
		groupID := ""
		buckets := make([]any, 0)
		for _, rawBucket := range getSlice(group["buckets"]) {
			bucket := getMap(rawBucket)
			id := cleanText(getString(bucket, "bucketId"))
			if strings.HasPrefix(id, "gemini-") {
				groupID = "gemini"
			} else if strings.HasPrefix(id, "3p-") {
				groupID = "third_party"
			}
			window := getString(bucket, "window")
			fraction, ok := getFloat(bucket["remainingFraction"])
			reset := getString(bucket, "resetTime")
			if !oneOf(window, "weekly", "5h") || !ok || reset == "" {
				continue
			}
			fraction = max(0, min(1, fraction))
			buckets = append(buckets, map[string]any{"id": id, "name": cleanText(firstString(bucket["displayName"], window)), "window": window, "remaining_fraction": fraction, "reset_at": reset})
		}
		if groupID != "" && len(buckets) > 0 {
			groups = append(groups, map[string]any{"id": groupID, "name": cleanText(firstString(group["displayName"], "Model group")), "buckets": buckets})
		}
	}
	if len(groups) == 0 {
		return nil, fmt.Errorf("Google returned no quota groups")
	}
	return normalizeQuotaSnapshot(map[string]any{"observed_at": isoTime(time.Now()), "tier": map[string]any{"id": tierID, "name": tierName}, "groups": groups}, getString(account, "email"))
}

func quotaAge(account Account, now time.Time) (time.Duration, bool) {
	snapshot := getMap(account["quota_snapshot"])
	observed, err := parseUTC(getString(snapshot, "observed_at"))
	if err != nil {
		return 0, false
	}
	age := now.Sub(observed)
	if age < 0 {
		age = 0
	}
	return age, true
}

func (q *QuotaService) Refresh(ctx context.Context, accounts *Accounts, force bool, progress quotaProgress) map[string]string {
	return q.refreshSelected(ctx, accounts, force, progress, "")
}

func (q *QuotaService) refreshSelected(ctx context.Context, accounts *Accounts, force bool, progress quotaProgress, selected string) map[string]string {
	now := time.Now().UTC()
	type item struct {
		email   string
		account Account
	}
	var fetch []item
	done := 0
	for _, email := range accounts.Order {
		if selected != "" && selected != email {
			continue
		}
		account := accounts.ByEmail[email]
		if age, ok := quotaAge(account, now); !force && ok && age < quotaCache {
			done++
			if progress != nil {
				progress(done, accounts.Len(), account, "cached")
			}
		} else {
			copyAccount := make(Account, len(account))
			for k, v := range account {
				copyAccount[k] = v
			}
			fetch = append(fetch, item{email, copyAccount})
		}
	}
	errorsByEmail := make(map[string]string)
	if len(fetch) == 0 {
		return errorsByEmail
	}
	workers := minInt(8, len(fetch))
	jobs := make(chan item)
	type result struct {
		item     item
		snapshot map[string]any
		err      error
	}
	results := make(chan result, len(fetch))
	var wg sync.WaitGroup
	for range workers {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for it := range jobs {
				snapshot, err := q.Fetch(ctx, it.account)
				results <- result{it, snapshot, err}
			}
		}()
	}
	go func() {
		for _, it := range fetch {
			jobs <- it
		}
		close(jobs)
		wg.Wait()
		close(results)
	}()
	changed := false
	var replaced []string
	for result := range results {
		done++
		state := "synced"
		original := accounts.ByEmail[result.item.email]
		if updated := getString(result.item.account, "token_data"); updated != "" && updated != getString(original, "token_data") {
			if q.vault != nil {
				app := &Application{vault: q.vault}
				old, ok := app.saveAccountSecret(ctx, result.item.account, updated)
				if !ok {
					errorsByEmail["storage"] = "OS vault unavailable; refreshed token stored in private accounts file"
				} else if old != "" {
					replaced = append(replaced, old)
				}
			} else {
				result.item.account["token_hash"] = hashToken(updated)
				rememberTokenExpiry(result.item.account, updated)
			}
			for _, key := range []string{"secret_ref", "token_data", "token_hash", "access_expires_at"} {
				if value, ok := result.item.account[key]; ok {
					original[key] = value
				} else {
					delete(original, key)
				}
			}
			changed = true
		}
		if result.err != nil {
			errorsByEmail[result.item.email] = result.err.Error()
			state = "failed"
		} else {
			original["quota_snapshot"] = result.snapshot
			changed = true
		}
		if progress != nil {
			progress(done, accounts.Len(), original, state)
		}
	}
	if changed {
		if err := q.store.Save(accounts); err != nil {
			errorsByEmail["store"] = err.Error()
		} else if q.vault != nil {
			(&Application{vault: q.vault}).deleteReplacedSecrets(ctx, replaced)
		}
	}
	return errorsByEmail
}

func tokenResetInfo(tokenData string) (string, bool) {
	inner := tokenObject(decodeToken(tokenData))
	if inner == nil {
		return "", false
	}
	expiry, ok := tokenExpiry(inner)
	if !ok {
		return "", false
	}
	return formatTokenReset(expiry, getString(inner, "refresh_token") != "")
}

func tokenResetFromExpiry(value string) (string, bool) {
	expiry, err := parseUTC(strings.TrimSpace(value))
	if err != nil || expiry.IsZero() {
		return "", false
	}
	return formatTokenReset(expiry, false)
}

func formatTokenReset(expiry time.Time, refreshAvailable bool) (string, bool) {
	if expiry.IsZero() {
		return "", false
	}
	diff := time.Until(expiry)
	if diff > 0 {
		mins := int(diff.Minutes())
		if mins >= 60 {
			return fmt.Sprintf("%dh %dm", mins/60, mins%60), true
		}
		return fmt.Sprintf("%dm %ds", mins, int(diff.Seconds())%60), true
	}
	if refreshAvailable {
		return "Access expired · refresh token available", true
	}
	return fmt.Sprintf("Expired %dm ago", int((-diff).Minutes())), true
}
